### ISSUE TITLE
Refactor the counter with generics

### ISSUE THREAD

**Issue by highcloud100** [yorkie-team/yorkie#765]
**What would you like to be added**:
---

Convert the raw types to generics in counter code.

https://github.com/yorkie-team/yorkie/blob/main/pkg/document/crdt/counter.go

**Why is this needed**:
---
Currently LongCnt, IntegerCnt types can be used as counter. 
So there is many duplicated code in counter. 
Moreover interface{} is used for supporting arbitrary type in crdt.Counter

https://github.com/yorkie-team/yorkie/blob/eaefcdc2ce6a1928ac3b2d58ccc9f60fcdd9d73a/pkg/document/crdt/counter.go#L53-L55

**Comment by gusah009** [comment]
@highcloud100 Hi, I’m interested in this issue. May I try it?

**Comment by highcloud100** [comment]
> @highcloud100 Hi, I’m interested in this issue. May I try it?

Yes, please !

**Comment by gusah009** [comment]
Hello, @highcloud100 @hackerwins

It’s been a while since I was assigned this issue, and I’m now working on resolving it. I have a question regarding whether making the counter generic would provide significant benefits.

First of all, I would like to mention that I am not very familiar with Go, so I am open to any corrections you may have.

I was considering implementing the counter using generics as shown below.

```go
// IntCounter is a counter that uses int32.
type IntCounter struct {
	value int32
}

// LongCounter is a counter that uses int64.
type LongCounter struct {
	value int64
}

type CounterType interface {
	IntCounter | LongCounter
}
```

By doing this, I was able to refactor the lengthy if branches in the Increase function into methods, effectively eliminating them. Additionally, I removed the code in NewCounter that checked and cast the type of Counter at runtime.

### Refactoring the Increase Method

**AS-IS**
```go
func (p *Counter) Increase(v *Primitive) (*Counter, error) {
	if !p.IsNumericType() || !v.IsNumericType() {
		return nil, ErrUnsupportedType
	}
	switch p.valueType {
	case IntegerCnt:
		intValue, err := castToInt(v.value)
		if err != nil {
			return nil, err
		}
		p.value = p.value.(int32) + intValue
	case LongCnt:
		longValue, err := castToLong(v.value)
		if err != nil {
			return nil, err
		}
		p.value = p.value.(int64) + longValue
	default:
		return nil, ErrUnsupportedType
	}

	return p, nil
}

// IsNumericType checks for numeric types.
func (p *Counter) IsNumericType() bool {
	t := p.valueType
	return t == IntegerCnt || t == LongCnt
}
```

**TO-BE**
```go
func (c *IntCounter) Increase(delta interface{}) error {
	deltaInt, err := castToInt(delta)
	if err != nil {
		return ErrUnsupportedType
	}
	if (deltaInt > 0 && c.value > math.MaxInt32-deltaInt) || (deltaInt < 0 && c.value < math.MinInt32-deltaInt) {
		return errors.New("int32 overflow")
	}
	c.value += deltaInt
	return nil
}

func (c *LongCounter) Increase(delta interface{}) error {
	deltaLong, err := castToLong(delta)
	if err != nil {
		return ErrUnsupportedType
	}
	if (deltaLong > 0 && c.value > math.MaxInt64-deltaLong) || (deltaLong < 0 && c.value < math.MinInt64-deltaLong) {
		return errors.New("int64 overflow")
	}
	c.value += deltaLong
	return nil
}
```

### Removing the NewCounter Function

**AS-IS**
```go
// NewCounter creates a new instance of Counter.
func NewCounter(valueType CounterType, value interface{}, createdAt *time.Ticket) (*Counter, error) {
	switch valueType {
	case IntegerCnt:
		intValue, err := castToInt(value)
		if err != nil {
			return nil, err
		}
		return &Counter{
			valueType: IntegerCnt,
			value:     intValue,
			createdAt: createdAt,
		}, nil
	case LongCnt:
		longValue, err := castToLong(value)
		if err != nil {
			return nil, err
		}
		return &Counter{
			valueType: LongCnt,
			value:     longValue,
			createdAt: createdAt,
		}, nil
	default:
		return nil, ErrUnsupportedType
	}
}
```

**TO-BE**
The function is removed. Instead, we can create a Counter like this:

```go
crdt.Counter{
	value:     counterValue,
	createdAt: createdAt,
}
```

### My Problem

However, I encountered a problem. In this function ([link to code](https://github.com/yorkie-team/yorkie/blob/0871fac98488c7de245f19a395c0be1d0317ac19/pkg/document/operations/increase.go#L46-L59)), we need to specify the type of crdt.Counter from the received Element. However, since Element does not use generics, there’s no way to determine the type at compile-time. As a result, we might have to perform runtime type checks and branch accordingly, like this:

```go
func (o *Increase) Execute(root *crdt.Root) error {
    parent := root.FindByCreatedAt(o.parentCreatedAt)
    switch cnt := parent.(type) {
    case *Counter[*IntCounter, int32]:
        value, ok := o.value.(*Primitive)
        if !ok {
            return errors.New("invalid value type")
        }
        delta, ok := value.Value().(int32)
        if !ok {
            return errors.New("invalid delta type")
        }
        if err := cnt.value.Increase(delta); err != nil {
            return err
        }
    case *Counter[*LongCounter, int64]:
        value, ok := o.value.(*Primitive)
        if !ok {
            return errors.New("invalid value type")
        }
        delta, ok := value.Value().(int64)
        if !ok {
            return errors.New("invalid delta type")
        }
        if err := cnt.value.Increase(delta); err != nil {
            return err
        }
    default:
        return ErrNotApplicableDataType
    }
    return nil
}
```
In that case, I’m not sure if using generics for this improvement is meaningful.

In conclusion, if only Counter among the many implementations of Element is made generic, we cannot resolve the branching issue, and I believe there’s a high possibility of introducing bugs during the refactoring process. What do you think?
